core/unix/rseq_linux.c

```diff
@@ -0,0 +1,377 @@
+ * is pre-empted within an rseq region, an abort handler is invoked.
+ * The feature is difficult to handle under binary instrumentation.
+ * We rely on the app following certain conventions, including containing a
+ * section holding a table of all rseq sequences.
+ */
+
+#include "../globals.h"
+#include "../module_shared.h"
+#include "module_private.h"
+#include "os_private.h"
+#include "rseq_linux.h"
+#include "../fragment.h"
+#ifdef HAVE_RSEQ
+#    include <linux/rseq.h>
+#else
+struct rseq_cs {
+    uint version;
+    uint flags;
+    uint64 start_ip;
+    uint64 post_commit_offset;
+    uint64 abort_ip;
+} __attribute__((aligned(4 * sizeof(uint64))));
+struct rseq {
+    uint cpu_id_start;
+    uint cpu_id;
+    uint64 ptr64;
+    uint flags;
+} __attribute__((aligned(4 * sizeof(uint64))));
+#    define RSEQ_FLAG_UNREGISTER 1
+#endif
+#include "include/syscall.h"
+#include <errno.h>
+
+vm_area_vector_t *d_r_rseq_areas;
+DECLARE_CXTSWPROT_VAR(static mutex_t rseq_trigger_lock,
+                      INIT_LOCK_FREE(rseq_trigger_lock));
+static volatile bool rseq_enabled;
+
+typedef struct _rseq_region_t {
+    app_pc start;
+    app_pc end;
+    app_pc handler;
+    /* XXX i#2350: Additional information will be added here for eliminating
+     * some existing limitations.
+     */
+} rseq_region_t;
+
+/* vmvector callbacks */
+static void
+rseq_area_free(void *data)
+{
+    HEAP_TYPE_FREE(GLOBAL_DCONTEXT, data, rseq_region_t, ACCT_VMAREAS, PROTECTED);
+}
+
+static void *
+rseq_area_dup(void *data)
+{
+    rseq_region_t *src = (rseq_region_t *)data;
+    rseq_region_t *dst =
+        HEAP_TYPE_ALLOC(GLOBAL_DCONTEXT, rseq_region_t, ACCT_VMAREAS, PROTECTED);
+    ASSERT(src != NULL);
+    *dst = *src;
+    return dst;
+}
+
+void
+d_r_rseq_init(void)
+{
+    VMVECTOR_ALLOC_VECTOR(d_r_rseq_areas, GLOBAL_DCONTEXT,
+                          VECTOR_SHARED | VECTOR_NEVER_MERGE, rseq_areas);
+    vmvector_set_callbacks(d_r_rseq_areas, rseq_area_free, rseq_area_dup, NULL, NULL);
+    if (rseq_is_registered_for_current_thread())
+        rseq_enabled = true;
+}
+
+void
+d_r_rseq_exit(void)
+{
+    vmvector_delete_vector(GLOBAL_DCONTEXT, d_r_rseq_areas);
+    DELETE_LOCK(rseq_trigger_lock);
+}
+
+bool
+rseq_get_region_info(app_pc pc, app_pc *start OUT, app_pc *end OUT, app_pc *handler OUT)
+{
+    rseq_region_t *info;
+    if (!vmvector_lookup_data(d_r_rseq_areas, pc, start, end, (void **)&info))
+        return false;
+    if (handler != NULL)
+        *handler = info->handler;
+    return true;
+}
+
+bool
+rseq_is_registered_for_current_thread(void)
+{
+    /* Unfortunately there's no way to query the current rseq struct.
+     * For 64-bit we can pass a kernel address and look for EFAULT
+     * vs EINVAL, but there is no kernel address for 32-bit.
+     * So we try to perform a legitimate registration.
+     */
+    struct rseq test_rseq = {};
+    int res = dynamorio_syscall(SYS_rseq, 4, &test_rseq, sizeof(test_rseq), 0, 0);
```

Comment:
Not sure if I'm missing something, but should this be RSEQ_FLAG_UNREGISTER, 0 instead of 0, 0?

Reply:
Nevermind, I think it's fine looking at the kernel [source](https://github.com/torvalds/linux/blob/d82991a8688ad128b46db1b42d5d84396487a508/kernel/rseq.c)

Having said that, the rseq syscall behavior, is this documented somewhere, or are we referring to the kernel's source, reverse engineering?

Reply:
As the comment tries to say, the syscall was poorly done, with no interface or documented method of querying. We have to resort to reverse engineering to find a hacky workaround.

Reply:
I should say, this is actually not very hacky: I'm thinking of my original plan of the EFAULT vs EINVAL which depends on the exact syscall ordering, but I abandoned that as the comment says b/c it doesn't work well in 32-bit.  Here we just try to register, and EINVAL vs success is part of the interface, so this does not seem so bad.

Reply:
Ok.